Add more context to `quit_if_file_exists` in `configure.py` & delete config.toml in CI

tgross35 opened the pull request:
If the `obj` directory isn't empty, the error message is subtle and not very helpful:

```
== clock drift check ==
  local time: Sun Jul  2 00:57:06 UTC 2023
  network time: Sun, 02 Jul 2023 00:57:06 GMT
== end clock drift check ==
sccache: Starting the server...
configure: error: Existing 'config.toml' detected.
== clock drift check ==
  local time: Sun Jul  2 00:57:06 UTC 2023
  network time: Sun, 02 Jul 2023 00:57:06 GMT
== end clock drift check ==
```

This makes it stand out and suggests how to resolve the issue:

```
== clock drift check ==
  local time: Sun Jul  2 02:11:30 UTC 2023
  network time: Sun, 02 Jul 2023 02:11:31 GMT
== end clock drift check ==
sccache: Starting the server...

configure: ERROR: Existing 'config.toml' detected. Exiting
Is objdir '/home/tmgross/projects/rust/obj' clean?

== clock drift check ==
  local time: Sun Jul  2 02:11:31 UTC 2023
  network time: Sun, 02 Jul 2023 02:11:31 GMT
== end clock drift check ==
```

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5341736522/jobs/9682843077) [(plain)](https://github.com/rust-lang/rust/commit/627ad4a9eb907e6d2aeea5ec92c15b21bb508f0c/checks/14459587953/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:c85c95e3d7251135ab7dc9ce3241c5835cc595a9)
Download action repository 'rust-lang/simpleinfra@master' (SHA:3999262f9f9933e66da0e5e168b90798c32a320e)
Download action repository 'actions/upload-artifact@v3' (SHA:0b7f8abb1508181956e8e162db84b466c27e18ce)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: mingw-check-tidy
---
Successfully tagged rust-ci:latest
Built container sha256:835d9253e364606498b7b985f454c218c3a482f0ca4462844ad1204497afe377
Uploading finished image to https://ci-caches.rust-lang.org/docker/6c79211b759db8881e4e083f32c23f7a9fde027e88ac4e65b52f7f5e0405fbb6bd8d3cd81afe7617920fe598db5be42068b29a40aff8b5cde467364f4587bbd8

<botocore.awsrequest.AWSRequest object at 0x7f6115525910>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
##[endgroup]
fmt check
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/src/ci/docker/README.md:44: trailing whitespace
some tidy checks failed
```

</details>

the8472 commented:
> The guide stops at building the image, but doesn't say anything about how to run it.

That's not correct, the `./src/ci/docker/run.sh x86_64-gnu` command in the previous section will _run_ the build/test job docker.

tgross35 commented:
Hm, it does not for me...

```
$ ./src/ci/docker/run.sh x86_64-gnu
Attempting with retry: docker build --rm -t rust-ci -f ...
[+] Building 139.4s (9/9) FINISHED                                                                             
...                                                                  0.0s
 => [internal] load metadata for docker.io/library/ubuntu:20.04                                           1.1s
 => [1/4] FROM docker.io/library/ubuntu:20.04@sha256:f8f658407c35733471596f25fdb4ed748b80e545ab57e84efbd  3.3s
...
 => [2/4] RUN apt-get update && apt-get install -y --no-install-recommends   g++   make   ninja-build   124.7s
 => [3/4] COPY scripts/sccache.sh /scripts/                                                               0.0s
 => [4/4] RUN sh /scripts/sccache.sh                                                                      3.7s
 => exporting to image                                                                                    6.5s
 => => exporting layers                                                                                   6.5s
 => => writing image sha256:ef3942f9859700ca0d7f35c69e76a2fc73097a63dd7fffc2d6a53fb6c1a3a09c              0.0s
 => => naming to docker.io/library/rust-ci                                                                0.0s
== clock drift check ==
  local time: Sun Jul  2 00:57:06 UTC 2023
  network time: Sun, 02 Jul 2023 00:57:06 GMT
== end clock drift check ==
sccache: Starting the server...
configure: error: Existing 'config.toml' detected.
== clock drift check ==
  local time: Sun Jul  2 00:57:06 UTC 2023
  network time: Sun, 02 Jul 2023 00:57:06 GMT
== end clock drift check ==
```

And that's it. I'll try to see what's going on with the script, maybe I'm just missing an environment variable or something

the8472 commented:
> configure: error: Existing 'config.toml' detected.

That's the container running but encountering an error.

tgross35 commented:
I'm not sure what it's talking about

```
$ fd config.toml
src/tools/error_index_generator/book_config.toml
```

Trying to chase it down

tgross35 committed:
Add more context to `quit_if_file_exists` in `configure.py`

Currently, having a dirty `obj/` directory is sufficient to abort CI
tests. This results in errors like the following:

```
...
== end clock drift check ==
sccache: Starting the server...
configure: error: Existing 'config.toml' detected.
== clock drift check ==
...
```

This is subtle and doesn't give a good idea as to what causes the issue.
With this patch, the error becomes more prominent and a resolution is
suggested:

```
== end clock drift check ==
sccache: Starting the server...

configure: ERROR: Existing 'config.toml' detected. Exiting
Is objdir '/home/tmgross/projects/rust/obj' clean?

== clock drift check ==
```

tgross35 commented:
It was just a dirty obj directory. The error message is kind of subtle and doesn't provide any context, I'm surprised more people don't hit this issue.

So I adjusted this PR to make the error stand out more and suggest checking that directory. Is that suitable, or would it be better to verify this directory doesn't exist in `run.sh`?

tgross35 commented:
@rustbot  ready

bors commented:
:umbrella: The latest upstream changes (presumably #112779) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Mark-Simulacrum commented:
This doesn't seem right to me, it should be fine to re-run a given docker image in the same directory without purging the cache. Maybe the previous error message for an existing config.toml is misleading and (for example) should be dropped if we're going to create the same file?

tgross35 commented:
Do you or others also run into this error if running twice? It’s a bit weird that this hasn’t been addressed in some way or another, unless there’s something about my setup that makes me hit it for some reason.

I can adjust this to just delete the config file and whatever else is needed to not conflict.

Mark-Simulacrum commented:
`./src/ci/docker/run.sh mingw-check-tidy` run twice for me seems to work without issues - so not sure if that command is special?

What command are you running?

tgross35 commented:
Same script but with x86_64-gnu, my output is in the comment above https://github.com/rust-lang/rust/pull/112916#issuecomment-1616248722

Mark-Simulacrum commented:
Ah, I see, I think I had an old branch. The error seems new as of https://github.com/rust-lang/rust/pull/110123 - I'm not sure that it should be an error, as the last comment on that PR suggests -- at minimum, we should silence that error during CI docker builds.

tgross35 commented:
Hm, yeah that is interesting to see the context. What do you think would be best? I think having a nicer error message is nice at a minimum, but I’ll update this PR to whatever makes the most sense.

cc @jyn514 since you were involved in the PR that Mark linked

jyn514 commented:
i think we should change the CI image to run `rm -f config.toml` before running configure. the issue here seems limited to CI and i don't think we should change the general behavior of configure.py.

jyn514 commented:
(i think my ideal behavior here would be "edit the file in place, keeping any existing settings, and only error if a setting conflicts", but that's too complicated to do in python and we can't move it to rust because it happens too early.)

Mark-Simulacrum commented:
I think it's also true that every other configure I've worked with has overwritten the configuration in place with no warning. I'm not convinced humans are often running configure and automation rarely wants to extend previous state I think.

Anyway, I'm fine with updating CI to delete configuration. That seems like a reasonable approach.

Mark-Simulacrum commented:
@tgross35 Can you update this PR to remove the file just before this line: https://github.com/rust-lang/rust/blob/master/src/ci/run.sh#L174 ? I think that's the right place for this delta to work best.

tgross35 commented:
Whoops forgot about this one, yes will do that change soon

tgross35 commented:
@Mark-Simulacrum this should be all set

@rustbot ready

Mark-Simulacrum commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 8a2022b108c520085fc486619ba2af39e35b1738 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 8a2022b108c520085fc486619ba2af39e35b1738 -->
<!-- homu: {"type":"Approved","sha":"8a2022b108c520085fc486619ba2af39e35b1738","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 8a2022b108c520085fc486619ba2af39e35b1738 with merge 2aae331706fee2a901766a0f0c008b263b077242...
<!-- homu: {"type":"BuildStarted","head_sha":"8a2022b108c520085fc486619ba2af39e35b1738","merge_sha":"2aae331706fee2a901766a0f0c008b263b077242"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5779994968/job/15664555418)
Approved by: Mark-Simulacrum
Pushing 2aae331706fee2a901766a0f0c008b263b077242 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5779994968/job/15664555418"},"merge_sha":"2aae331706fee2a901766a0f0c008b263b077242"} -->

rust-timer commented:
Finished benchmarking commit (2aae331706fee2a901766a0f0c008b263b077242): [comparison URL](https://perf.rust-lang.org/compare.html?start=72c6b8d36fa27207fb66b2d8386ad959aa10e6bb&end=2aae331706fee2a901766a0f0c008b263b077242&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=72c6b8d36fa27207fb66b2d8386ad959aa10e6bb&end=2aae331706fee2a901766a0f0c008b263b077242&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6% | [0.6%, 0.6%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.6% | [0.6%, 0.6%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=72c6b8d36fa27207fb66b2d8386ad959aa10e6bb&end=2aae331706fee2a901766a0f0c008b263b077242&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.5% | [-2.8%, -2.1%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=72c6b8d36fa27207fb66b2d8386ad959aa10e6bb&end=2aae331706fee2a901766a0f0c008b263b077242&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-0.9%, -0.9%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.3% | [-2.3%, -2.3%] | 1     |
| All ❌✅ (primary)                 | -0.9% | [-0.9%, -0.9%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=72c6b8d36fa27207fb66b2d8386ad959aa10e6bb&end=2aae331706fee2a901766a0f0c008b263b077242&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 650.942s -> 650.283s (-0.10%)
